hs-ucan/library/Crypto/Key/Asymmetric/Public/Types.hs

```diff
@@ -1,70 +1,83 @@
+import qualified Data.Aeson.Types                                 as Json
 import           RIO
-import qualified RIO.Text                                      as Text
+import qualified RIO.ByteString.Lazy                              as Lazy
+import qualified RIO.Text                                         as Text
 import           Servant.API
 import           Test.QuickCheck
 
-import           Web.UCAN.Internal.RSA2048.Pair.Types          as Pair
+import           Web.UCAN.Internal.RSA2048.Pair.Types             as Pair
 
-import           Web.UCAN.Internal.Orphanage.Ed25519.PublicKey ()
-import           Web.UCAN.Internal.Orphanage.RSA2048.Public    ()
+import           Web.UCAN.Internal.Orphanage.Ed25519.PublicKey    ()
+import           Web.UCAN.Internal.Orphanage.RSA2048.Public       ()
+import           Web.UCAN.Internal.Orphanage.Secp256k1.PublicKey  ()
 
 data Public
-  = Ed25519PublicKey Crypto.Ed25519.PublicKey
-  | RSAPublicKey     Crypto.RSA.PublicKey
-  deriving Eq
+  = Ed25519PublicKey   Crypto.Ed25519.PublicKey
+  | RSAPublicKey       Crypto.RSA.PublicKey
+  | Secp256k1PublicKey Crypto.Secp256k1.PubKeyXY
+  deriving (Eq)
 
 instance Show Public where
   show = \case
-    Ed25519PublicKey ed -> Text.unpack $ textDisplay ed
-    RSAPublicKey     pk -> show pk
+    Ed25519PublicKey   ed -> Text.unpack $ textDisplay ed
+    RSAPublicKey       pk -> show pk
+    Secp256k1PublicKey pk -> Text.unpack $ textDisplay pk
 
 instance Display Public where
-  textDisplay (Ed25519PublicKey pk) = textDisplay pk
-  textDisplay (RSAPublicKey     pk) = textDisplay pk
+  textDisplay (Ed25519PublicKey   pk) = textDisplay pk
+  textDisplay (RSAPublicKey       pk) = textDisplay pk
+  textDisplay (Secp256k1PublicKey pk) = textDisplay pk
 
 instance Arbitrary Public where
   arbitrary = oneof
     [ Ed25519PublicKey <$> arbitrary
     , RSAPublicKey . Pair.pk <$> arbitrary
+    , Secp256k1PublicKey <$> arbitrary
     ]
 
 instance ToHttpApiData Public where
   toUrlPiece = textDisplay
 
 instance FromHttpApiData Public where
-  parseUrlPiece txt =
-    if "MII" `Text.isPrefixOf` txt
-      then RSAPublicKey     <$> parseUrlPiece txt
-      else Ed25519PublicKey <$> parseUrlPiece txt
+  parseUrlPiece text =
+    case (Json.decode . Lazy.fromStrict . Text.encodeUtf8 $ text) :: Maybe Public of
+      Just pub -> Right pub
+      Nothing -> Left "Failed to decode the JSON formatted public key"
 
 instance IsString (Either Text Public) where
   fromString = parseUrlPiece . Text.pack
 
 instance FromJSON Public where
-  parseJSON = withText "PublicKey" \txt ->
-    case parseUrlPiece txt of
-      Right pk -> return pk
-      Left msg -> fail $ Text.unpack msg
+  parseJSON =
+    withObject "PublicKey" \j ->
+      ((,) <$> j .: "type" <*> j .: "key") >>= \(t, k) ->
+      case fromConstructor t k of
+        Right pub -> return pub
+        Left err -> Json.parseFail (Text.unpack err)
```

Comment:
Same question here as with `parseUrlPiece`, any other implications?

Reply:
Hm. Is it possible that the route `/v2/api/user/did` uses this? The servant request type there is
```hs
type SetAuthenticated
  =  Summary "Update Public Key"
  :> Description "Set currently authenticated user's root public key to another one"
  --
  :> ReqBody '[JSON] Key.Public
  --
  :> Auth.HigherOrder
  :> PutNoContent
```
And I'm thinking the `ReqBody '[JSON] Key.Public` probably uses this instance for going JSON -> Public Key.

Reply:
Same goes for the `SetViaChallenge` route. That one is actually used in the recovery process in the dashboard (I'm not aware of any production usage of `SetAuthenticated`).

Reply:
Thanks! I think it makes sense here to use the new structure `{ type: ..., key: ... }`, for both of these. But I guess I could use the old logic as a fallback, in case just a string was passed? What do you think?

Reply:
I guess that'd work.
Here's two more options:
- Full blown: Create a `v3` route that has the new behavior, keep the old route with the old behavior by creating a `Key.Public.Oldstyle` type and implement the old `FromJSON` for it.
- No backwards compatibility at all: We figure out all places that use this function. If it turns out that's pretty much only the dashboard app's recovery process, then we just deploy a fix roughly at the same time as we're updating our server.

Reply:
I've implemented the V3 routes 👍